### DISCUSSION TITLE
[question] is is possible to format code using tree-sitter?

### DISCUSSION THREAD

**Discussion by hronro** [tree-sitter/tree-sitter#3107] [Libraries & tools] [answered]
### Problem

This is probably a silly question since I don't know much about tree-sitter internals.

For me, syntax highlighting and code formatting are must have features in a modern code editor.

I know some languages come with an official code formatter (go, rust...), and some languages have a community code formatter (like prettier for js/ts), but they often requires additional softwares to be installed or the performance of the formatter is not ideal.

Given that Tree-sitter is primarily designed for code editors and has already handled the necessary AST parsing, could we utilize Tree-sitter (or at least its parsed AST) for code formatting?

**Comment by clason** [comment]
Yes.

**Comment by hronro** [comment]
@clason Could you provide more details on this? Are you saying that the current version of tree-sitter can already format code (even though I haven't found any code editor integrated with tree-sitter that offers this feature), or are some changes needed in tree-sitter to enable code formatting?

**Reply by clason** [reply]
Neovim does it; see https://github.com/nvim-treesitter/nvim-treesitter/blob/master/scripts/format-queries.lua. Of course, some assembly required.

**Comment by sogaiu** [comment, marked as answer]
[Possibly of interest](https://github.com/tweag/topiary)
